### PR TITLE
PVC mount support

### LINKED ISSUE #182: How should Krustlet handle the Container Storage Interface?
Many of the cloud provider volumes have Container Storage Interface (CSI) plugin implementations to handle mounting volumes to containers. Right now, the attachdetach-controller will attach cloud provider volumes to VMs but will not actually mount the volume. So a few of questions for discussion:

1. Should we implement CSI calls inside of krustlet itself (or at least something similar)?
2. Will it be possible to reuse the current Go plugins or are they too container-centric?
3. Would it be better to leave volume mounting up to individual implementations?
4. Would a good middle ground just be to implement drivers for open source file mounting (NFS, SMB, etc)?

For reference, the CSI docs can be found here: https://kubernetes-csi.github.io/docs/

### PR DESCRIPTION
**Pull request by bacongobbler**
Adds support for mounting PersistentVolumeClaims to a Pod via the Container Storage Interface.

Closes #182

### CODE REVIEW

**Comment by bacongobbler** [comment]
I got everything working here according to the CSI examples. I DO NOT expect anyone else to be able to reproduce the same environment, but if you want to try giving it a whirl... Here goes:

1. compile the following projects from HEAD, and move them into your PATH: 
   a. https://github.com/kubernetes-csi/csi-driver-host-path
   b. https://github.com/kubernetes-csi/node-driver-registrar
   c. https://github.com/kubernetes-csi/external-provisioner

2. In separate terminals spawn the following:

```
# hostpathplugin will need root permissions to execute 'mount'
./hostpathplugin --nodeid ultralisk --endpoint unix://csi/csi.sock -v 5
./csi-node-driver-registrar --csi-address /csi/csi.sock --kubelet-registration-path /csi/csi.sock --plugin-registration-path ~/.krustlet/plugins
./csi-provisioner --csi-address /csi/csi.sock --kubeconfig ~/.kube/config
```

3. Start Krustlet. Approve the bootstrap TLS certificate.
4. Create the storage class: `kubectl create -f https://raw.githubusercontent.com/kubernetes-csi/csi-driver-host-path/master/examples/csi-storageclass.yaml`
5. Create the PVC: `kubectl create -f https://raw.githubusercontent.com/kubernetes-csi/csi-driver-host-path/master/examples/csi-pvc.yaml`
6. Create the modified `k8s.yaml` example in `demos/wasi`

**Comment by bacongobbler** [comment]
Note: this is blocked until https://github.com/kflansburg/k8s-csi/pull/1 is merged and a new version of k8s-csi is released, as it will not compile until it's been re-compiled against tonic 0.3. I've pointed Krustlet at a temporary fork for the time being.

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (thomastaylor312) [inline crates/kubelet/src/volume/persistentvolumeclaim.rs:290]: Can you add more context here?

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (thomastaylor312) [inline crates/kubelet/src/volume/persistentvolumeclaim.rs:317]: Same with more context here

File: crates/kubelet/src/volume/secret.rs
Comment (thomastaylor312) [inline crates/kubelet/src/volume/secret.rs:15]: Nit: I know this is copied code, but if you feel so inclined, could you change this to `into_iter`? We are consuming the data anyway

File: demos/wasi/hello-world-rust/k8s.yaml
Comment (thomastaylor312) [inline demos/wasi/hello-world-rust/k8s.yaml:41]: We should probably do a separate example using the same hello world but with PVC

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
This is looking pretty good! We definitely should document all the steps required (basically at least the components needed) and a bug "THIS IS VERY ALPHA" label somewhere

File: crates/kubelet/src/provider/mod.rs
Comment (thomastaylor312) [reply crates/kubelet/src/provider/mod.rs:132]: After looking at the rest of the code, I'm downgrading this to a nit

File: crates/kubelet/src/volume/mod.rs
Comment (kflansburg) [inline crates/kubelet/src/volume/mod.rs:177]: Should fetching things from the API be moved into the respective modules?

File: crates/kubelet/src/volume/mod.rs
Comment (kflansburg) [inline crates/kubelet/src/volume/mod.rs:219]: Isnt this handled by k8s when the PVC is bound?

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (kflansburg) [inline crates/kubelet/src/volume/persistentvolumeclaim.rs:79]: Is it possible for these to be created with the other values? Is there value in creating these new types rather than just doing `anyhow::bail` with some context?

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (kflansburg) [inline crates/kubelet/src/volume/persistentvolumeclaim.rs:147]: This looks like copypasta

**Review by kflansburg — COMMENTED** [review]
I don't understand this enough to approve but had some comments.

**Review by itowlson — CHANGES_REQUESTED** [review]
I have basically no idea whether this is right or wrong, but I sure do have a lot of opinions about how to organise code I don't understand!

File: crates/kubelet/src/volume/mod.rs
Comment (thomastaylor312) [reply crates/kubelet/src/volume/mod.rs:189]: Most likely, but that could be captured as another issue (we should improve logging in a couple places

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (thomastaylor312) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:185]: You have to specify a default storage class at the cluster level, so the admission controller will handle it. But I actually think we'd need this error in the case we do get an empty storage class name

File: crates/kubelet/src/provider/mod.rs
Comment (bacongobbler) [reply crates/kubelet/src/provider/mod.rs:132]: I like that idea @kflansburg.

I'm curious how we should handle cases where a PVC is requested but the provider doesn't support CSI plugins though. Should we error out and say that the pod cannot be scheduled? Trying to determine what is the core criteria for being a Provider here.

File: crates/kubelet/src/volume/mod.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/mod.rs:189]: Oh these are just your typical garden variety debug messages. I'll remove them

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:161]: Yes. This validation is performed by the Kubernetes API, so we need to validate and sanitize this metadata as well.

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:225]: Yes absolutely.

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:260]: Somehow I missed this but yeah doing a `.join` would be muuuuuch better.

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:251]: Not really, no. We could if we implemented the validation logic as above but at this point the call to `unwrap()` is totally unsafe. I'll see what I can do about this.

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:264]: We need this for the call to `publish` as well for some reason... Dunno why but it's a required field in the gRPC call.

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:290]: garden variety debug statement; will remove

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:317]: garden variety debug statement; will remove

File: crates/kubelet/src/volume/secret.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/secret.rs:15]: yep!

File: demos/wasi/hello-world-rust/k8s.yaml
Comment (bacongobbler) [reply demos/wasi/hello-world-rust/k8s.yaml:41]: agreed. Might need to point to the documentation on setting up the example too, but yes.

File: crates/kubelet/Cargo.toml
Comment (bacongobbler) [reply crates/kubelet/Cargo.toml:53]: yes!

File: crates/kubelet/src/volume/mod.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/mod.rs:219]: If I can recall from memory... The volume manager code selects and determines the correct storage class. This is performed at runtime. Kubernetes does not handle this for the Kubelet.

I'll see if I can find the relevant code bits again and drop a doc comment to give some more context around this.

File: crates/kubelet/src/volume/mod.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/mod.rs:27]: oh I totally agree here.

File: crates/kubelet/src/provider/mod.rs
Comment (thomastaylor312) [reply crates/kubelet/src/provider/mod.rs:132]: Yeah, I think we'd just error the pod in a case where pvcs are not supported and one is requested

File: crates/kubelet/src/volume/mod.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/mod.rs:27]: Though to be fair in the future I am planning on refactoring all of this out as part of #482. I still think this is a great idea in the interim however

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:85]: Is there a good example of this somewhere I can use as a reference guide?

File: crates/kubelet/src/provider/mod.rs
Comment (kflansburg) [reply crates/kubelet/src/provider/mod.rs:132]: Yeah, I believe there are some well known errors / behavior here, you get a reason `FailedMount` and an error like `Failed to mount x for pod: CSI driver support not implemented.`. The only tricky part I would think is how to interrupt the state machine from where this is called. Not sure about that.

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (thomastaylor312) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:85]: Take a look at the [bindle errors](https://github.com/deislabs/bindle/blob/b51ae14f26be2a647ceffc38ace26f3290dfacae/src/provider/mod.rs#L119) for an example

Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:85]: Thank you! I'll have a look at this today.

File: crates/kubelet/src/volume/mod.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/mod.rs:177]: I'd prefer that the functions deal with the concrete type rather than all the kube client fetching and whatnot. That way the transition from this design to #482 will be MUCH easier.

That being said let's go forward with this design for now and see if we need to revisit it after we start tackling #482. Either way this is all private functions so it doesn't really affect the public API at this point :)

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:113]: Is there something I should address here?

File: crates/kubelet/src/volume/configmap.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/configmap.rs:34]: I do agree with your comment here. While I didn't write this code as part of the PR, I'll see what I can do to improve the code's readability.

Comment (bacongobbler) [reply crates/kubelet/src/volume/configmap.rs:34]: I tried to tackle this, but my failure to understand generics fails me once again. `configmap.binary_data` returns a `BTreeMap<String, ByteString>` whereas `configmap.data` returns a `BTreeMap<String, String>`. So a function signature like `async fn write_all(data: BTreeMap<String, String>) -> anyhow::Result<()>` will not work in this context.

Any ideas?

File: crates/kubelet/src/volume/mod.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/mod.rs]: hmm... @kflansburg raised this concern in another comment as well.

I do agree with both your feedback. However, I do intend to refactor this some more in the next release (#482 should help a lot with this). Is it all right if we move forward here without addressing this feedback at this time, or would you prefer that I address it now before we merge?

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:116]: For context, yes this is copypasta from the rust docs. Reference implementation: https://doc.rust-lang.org/std/str/trait.FromStr.html

File: crates/kubelet/src/volume/configmap.rs
Comment (itowlson) [reply crates/kubelet/src/volume/configmap.rs:34]: `async fn write_all<T,C: AsRef<[u8]> + Unpin>(data: BTreeMap<String, T>, accessor: (&T -> C)) -> anyhow::Result<()>` maybe?  The `C` is the input type to `tokio::fs::write`; the `T` is the type of the k8s configmap entry values; and `acccessor` is the extractor function that takes you from T to C (the identity for String, the `.0` accessor for ByteString).

If it doesn't fall out fairly easily then I would certainly punt on it - it's not a biggie.

File: crates/kubelet/src/volume/mod.rs
Comment (itowlson) [reply crates/kubelet/src/volume/mod.rs]: Ship it as is :shipit: and I'll just watch #482 like a hawk 🐦 and yes I am aware that looks more like a parrot

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (itowlson) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:116]: I think the issue might be `s/reclaim policy/access mode`?

Comment (kflansburg) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:116]: Yep, it seemed like it was from another struct.

Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:116]: :headsmack: I see that now. Will address tomorrow :)

File: crates/kubelet/src/volume/configmap.rs
Comment (thomastaylor312) [reply crates/kubelet/src/volume/configmap.rs:34]: I actually think you could just do `async fn write_all<T: AsRef<[u8]>>(data: BTreeMap<String, T>)` as both `String` and `ByteString` implement `AsRef<[u8]>`

File: crates/kubelet/src/volume/configmap.rs
Comment (thomastaylor312) [inline crates/kubelet/src/volume/configmap.rs:14]: Nit: We are consuming (and moving) the data here, so `into_iter` is a better fit

File: crates/kubelet/src/volume/mod.rs
Comment (thomastaylor312) [inline crates/kubelet/src/volume/mod.rs:114]: I missed this the first time. I actually don't think we can let this clean up the persistent volume because otherwise it will be removing the data from the PV. The PV needs to be unmounted before we clean up the directory. I am guessing the volume mounter takes care of this cleanup and even if it doesn't, we can make it do it later when we try to port it to Rust. For now I am fine leaving an orphaned directory if that is the case

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (thomastaylor312) [inline crates/kubelet/src/volume/persistentvolumeclaim.rs:295]: Are we supposed to be handling the `NodeUnpublishVolume` request on pod termination? I don't think we are doing that in the providers and so we are leaving volumes published to a node when a pod is no longer there, which will cause errors when they get reused

File: docs/howto/csi.md
Comment (thomastaylor312) [inline docs/howto/csi.md:23]: Only external provisioner will be I think. The registrar is something that should be built into CSI implementations or just leverage the existing project

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
Docs look really good. I realized we were missing some things this second go around

File: crates/kubelet/src/volume/mod.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/mod.rs:114]: Fair enough. And I think leaving an orphaned directory is fine since it's a `mount`/`umount` anyways

File: docs/howto/csi.md
Comment (thomastaylor312) [reply docs/howto/csi.md:23]: So after investigating more, I don't think we have any immediate plans to port these over. Looks like external-provisioner is not linux specific. The node-driver-registrar is OS specific and we probably need to contribute back to it so it will work on Mac (which is a really easy fix)

Comment (thomastaylor312) [reply docs/howto/csi.md:23]: Opened a PR to allow for building on Darwin: https://github.com/kubernetes-csi/node-driver-registrar/pull/133

File: crates/kubelet/src/volume/configmap.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/configmap.rs:34]: I'm having a really tough time cranking this one out in a timely period. The function signatures are good, but it's the return type on the closure that's throwing me a curve ball.

This is what it looks like right now:

```
async fn write_all<T,F>(data: BTreeMap<String, T>, map_fn: F) -> anyhow::Result<()>
where
    T: AsRef<[u8]> + Unpin,
    F: FnMut(String, T) -> Map<Iter<String, T>, ???>,
{}
```

Reason being is that it's a closure consuming `items` and `path`, so you have to pass in a closure. And I need to take a deeper dive at how `map` actually works to determine the correct return type for F.

I'll file an issue for now so we can track this for later.

Comment (bacongobbler) [reply crates/kubelet/src/volume/configmap.rs:34]: https://github.com/deislabs/krustlet/issues/503

Comment (itowlson) [reply crates/kubelet/src/volume/configmap.rs:34]: No worries.  Sorry for sending you down a dead end.

File: crates/kubelet/src/volume/persistentvolumeclaim.rs
Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:295]: Yeah... I think we need that call. Where would be the best place to implement that call?

Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:295]: Looking into implementing this in the `Terminated` state, just before calling `state_reader.stop()`.

Comment (thomastaylor312) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:295]: @bacongobbler I was actually thinking it might be better to hook something in when the state machine terminates or to introduce a new common state called `volume cleanup`

Comment (kflansburg) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:295]: IMO we should try not to add this to the state machine. It would be preferable that providers which do not use the common states do not have to know / remember to re-implement this.

Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:344]: I hooked this into the `Terminated` state for the time being. Love you both but I cannot tell which direction y'all want me to take 😂 

Can you two please converse and let me know what your decision is after working through the design?

The nice thing is that I was able to refactor a bit of the volume package. Most of it should easily hook into whichever system y'all want me to write.

Comment (thomastaylor312) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:330]: After talking this over offline, we are going to stick with having the cleanup be in the `Terminated` state for now. However, we realized that Krator needs some sort of a deregistration hook and that has been captured in #504

Comment (bacongobbler) [reply crates/kubelet/src/volume/persistentvolumeclaim.rs:312]: Okay! I added a note in the code along with a link to the ticket. Thanks everyone.

**Comment by thomastaylor312** [comment]
Looks like the doc tests are failing @bacongobbler. I also am seeing a lot of clippy errors, but I can fix those later

**Comment by bacongobbler** [comment]
Oh yeah I see that. Never run the doc tests locally... Will fix those right up!

Thanks for catching the clippy errors as well.

**Comment by thomastaylor312** [comment]
One last doc add I just thought of. In some clusters, the `system:nodes` group doesn't have access to read `StorageClass` objects. So if the Pod is erroring saying that the kubelet doesn't have access, you'll have to create the following cluster role and role binding to allow the nodes group to access them:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: storageclass-reader
rules:
  - apiGroups: ["storage.k8s.io"]
    resources: ["storageclasses"]
    verbs: ["get", "watch", "list"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: node-storageclass-reader
subjects:
  - kind: Group
    name: system:nodes
    apiGroup: rbac.authorization.k8s.io
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: storageclass-reader
```

**Comment by bacongobbler** [comment]
Good call. I just added it to the HOWTO doc as an addendum. Let me know if you'd like me to make further changes/tweaks there.

**Review by thomastaylor312 — APPROVED** [review]

**Comment by bacongobbler** [comment]
Thanks @thomastaylor312!

@itowlson are there any further changes you'd like me to address here?

**Review by itowlson — APPROVED** [review]
LGTM

**Pull request merged**
